src/GafferImage/ImageStats.cpp

```diff
@@ -198,116 +236,222 @@ const DeepState *ImageStats::deepState() const
 	{
-		static_cast<FloatPlug *>( output )->setValue( 0.0f );
-		return;
+		ImagePlug::GlobalScope s( context );
+		const Imath::Box2i area = areaPlug()->getValue();
+		const Imath::Box2i dataWindow = flattenedInPlug()->dataWindowPlug()->getValue();
+		boundsIntersection = BufferAlgo::intersection( area, dataWindow );
+		areaMult = double(area.size().x) * area.size().y;
 	}
 
-	// Loop over the ROI and compute the min, max and average channel values and then set our outputs.
-	Sampler s( inPlug(), channelName, area );
+	if( output == tileStatsPlug() )
+	{
+		Imath::V2i tileOrigin = context->get<Imath::V2i>( ImagePlug::tileOriginContextName );
+		const Imath::Box2i tileBound = BufferAlgo::intersection(
+			Imath::Box2i( boundsIntersection.min - tileOrigin, boundsIntersection.max - tileOrigin ),
+			Imath::Box2i( Imath::V2i( 0 ), Imath::V2i( ImagePlug::tileSize() ) )
+		);
+
+		IECore::ConstFloatVectorDataPtr channelData = flattenedInPlug()->channelDataPlug()->getValue();
 
-	float min = Imath::limits<float>::max();
-	float max = Imath::limits<float>::min();
-	double sum = 0.;
+		float min = Imath::limits<float>::max();
+		float max = Imath::limits<float>::min();
+		double sum = 0.;
 
-	for( int y = area.min.y; y < area.max.y; ++y )
-	{
-		for( int x = area.min.x; x < area.max.x; ++x )
+		const std::vector<float> &channel = channelData->readable();
+		for( int y = tileBound.min.y; y < tileBound.max.y; ++y )
 		{
-			float v = s.sample( x, y );
-			min = std::min( v, min );
-			max = std::max( v, max );
-			sum += v;
+			for( int x = tileBound.min.x; x < tileBound.max.x; ++x )
+			{
+				float v = channel[ x + y * ImagePlug::tileSize() ];
+				min = std::min( v, min );
+				max = std::max( v, max );
+				sum += v;
+			}
 		}
-	}
 
-	if( output->parent<Plug>() == minPlug() )
-	{
-		static_cast<FloatPlug *>( output )->setValue( min );
+		static_cast<ObjectPlug *>( output )->setValue( new IECore::V3dData( Imath::V3d( min, max, sum ) ) );
 	}
-	else if( output->parent<Plug>() == maxPlug() )
+	else if( output == allStatsPlug() )
 	{
-		static_cast<FloatPlug *>( output )->setValue( max );
-	}
-	else if( output->parent<Plug>() == averagePlug() )
-	{
-		static_cast<FloatPlug *>( output )->setValue(
-			sum / double( (area.size().x) * (area.size().y) )
+		if( BufferAlgo::empty( boundsIntersection ) )
+		{
+			static_cast<ObjectPlug *>( output )->setValue( new IECore::V3dData( Imath::V3d( 0 ) ) );
+			return;
+		}
+		float min = Imath::limits<float>::max();
+		float max = Imath::limits<float>::min();
+		double sum = 0.;
+
+		// We traverse in TopToBottom order because floating point precision means that changing
+		// the order to sum in could produce slightly non-deterministic results
+		ImageAlgo::parallelGatherTiles(
+			flattenedInPlug(),
+			// Tile
+			[this] ( const ImagePlug *imageP, const Imath::V2i &tileOrigin )
+			{
+				return boost::static_pointer_cast<const IECore::V3dData>( tileStatsPlug()->getValue() )->readable();
```

Comment:
I'm not entirely sure how this interacts with the result passing in the TBB pipeline, but this is returning a reference to data whose ownership is dropped as soon as we return. If the TBB pipeline doesn't take a copy (and I don't _think_ it will), we can get crashes if the value is removed from the cache before the gather completes. To avoid doubt, let's either explicitly pass V3d or V3dDataPtr by value.

Reply:
I guess the other factor is the rules for automatic deduction of lambda return types, but since I don't know those off the top of my head either, I think something explicit would be preferable.